src/compile/data/bin.ts

```diff
@@ -19,36 +19,42 @@ function numberFormatExpr(expr: string, format: string) {
 function parse(model: Model): Dict<VgTransform[]> {
   return model.reduceFieldDef(function(binComponent: Dict<VgTransform[]>, fieldDef: FieldDef, channel: Channel) {
     const bin = model.fieldDef(channel).bin;
+    // Make this bin variable always an object
+    // and extend it with maxbins if needed here.
+    let transform: VgTransform[] = [];
     if (bin) {
-
-      let binTrans: VgTransform = extend({
-        type: 'bin',
-        field: fieldDef.field,
-        as: [field(fieldDef, {binSuffix: 'start'}), field(fieldDef, {binSuffix: 'end'})]
-      },
-        // if bin is an object, load parameter here!
-        typeof bin === 'boolean' ? {} : bin
-      );
-
-      const transform: VgTransform[] = [];
-      if (!binTrans.extent) {
-        const extentSignal = model.getName(fieldDef.field + '_extent');
-        transform.push({
-          type: 'extent',
+      const key = hash(bin) + fieldDef.field;
+      // FIXME: Check
+      if (!binComponent[key]) {
+        let binTrans: VgTransform = {
```

Comment:
Currently this is  extended with `extent` below.

Also, this spread code is poorly formatted. 

This is way better:

```
let binTrans: VgTransform = {
   type: bin, 
   field: fieldDef.field,
   as: [field(fieldDef, {binSuffix: 'start'}), field(fieldDef, {binSuffix: 'end'})], 
   ...(isBoolean(bin) ? {} : bin}
}
```

You can also make it const 

```
let binTrans: VgTransform = {
   type: bin, 
   field: fieldDef.field,
   as: [field(fieldDef, {binSuffix: 'start'}), field(fieldDef, {binSuffix: 'end'})], 
   ...(isBoolean(bin) ? {} : bin},
   // add extent if it's not specified
   ...(isBoolean(bin) || bin.extent === undefined ? {} : {extent: {signal: extentSignal}})
}
```

This would require minor code reordering and revision (hint: revise `if (!binTrans.extent) {` below), but this should be very simple to do.  I'm gonna leave that for you to figure out as it should be a good exercise for you.